serve/mlc_serve/engine/staging_engine_worker.py

```diff
@@ -114,8 +112,13 @@ def add(self, request_states: list[RequestState]):
                 if (
                     request_state.validation_err is not None
                     or request_state.prompt_len >= self.max_context_length
+                    or request_state.prompt_len >= self.max_num_batched_tokens > 0
+                    or (self.cache_manager.get_kv_cache_size() - request_state.prompt_len) < 0
+                    or (self.cache_manager.get_kv_cache_size() - request_state.prompt_len) / (len(self.current_batch) + 1) < self.max_decode_steps
```

Comment:
Again, assuming that the cache eviction is working, I don't understand why we need this condition.

Reply:
for example, previously we had `prompt_allocate_ratio == 1.5`, `max_num_batched_tokens == 200`, `cache size == 250` and you get a request with prompt == 200. It passes all your verification, request is added into list to be processed, but in the `_adjust_batch` you had a verification that prompt_len*prompt_allocate_ratio not less than cache size. But for above parameters, such verification failed and we received infinite attempts to add and each time this attempt fails. This is a reason of the hang. 

I substituted verification in the _adjust_batch vs `prompt_allocate_ratio` by verification of the enough room for each request - `self.cache_manager.get_max_new_tokens() <= self.max_decode_steps` but I still need to verify if we have an ability to add a request for inference and get rid of it on the submissin stage to avoid infinit attempts in `_adjust_batch`. This is a reason of the second condition in this place.

Reply:
Also I realized that both the existing and the new conditions are suboptimal for Mistral since with SWA not all prompt tokens need to have its KV values cached. We implemented such optimization in https://github.com/mlc-ai/mlc-llm/blob/main/mlc_llm/relax_model/llama_batched_vllm.py#L89-L92

So for example, 

```
if (
    self.cache_manager.get_free_space()
    <= self.prompt_allocate_ratio * num_tokens
):
```
can be optimized to

```
if (
    self.cache_manager.get_free_space()
    <= self.prompt_allocate_ratio * min(num_tokens, window_size)
):
```

Reply:
I do not see reason for leaving of `prompt_allocate_ratio`, as I pointed it can be easily broken by the next request on the next iteration of the loop. So, if we have situation which breaks this condition, it's no sense to have it.

As for sliding window - seems yes, it should be improved in two places 1) for requests which is preempted from kv cache should be cut if len(prompt) exceed `max_num_batched_tokens` but not from begin, but from the end of token list, right? 2) in case of new comparison instead of `prompt_allocate_ratio`, we need to take into account `window_size` as well.

Reply:
1) Your sentence is not making a sense, but I guess yes, we only need to cache the most recent token `window_size` tokens.

Reply:
WSA will be addressed in separate request